torch/csrc/jit/type.h

```diff
@@ -452,6 +453,54 @@ struct TORCH_API ListType : public Type {
   bool has_free_variables_;
 };
 
+struct FutureType;
+using FutureTypePtr = std::shared_ptr<FutureType>;
+
+struct TORCH_API FutureType : public Type {
+  // It's not exactly a singleton, but there should be exactly once instance of
+  // Future[T] for every T
+  static constexpr bool is_singleton = true;
+  friend struct Type;
+  template<typename ... T>
+  static FutureTypePtr create(TypePtr elem) {
+    return FutureTypePtr(new FutureType(std::move(elem))); // NOLINT(modernize-make-shared)
+  }
+  bool operator==(const Type& rhs) const override {
+    if (auto rhs_ = rhs.cast<FutureType>()) {
+      return *getElementType() == *rhs_->getElementType();
+    }
+    return false;
+  }
+  bool requires_grad() const override {
+    return elem->requires_grad();
+  }
+  std::string str() const override {
+    std::stringstream ss;
+    ss << "Future[" << getElementType()->str() << "]";
```

Comment:
nit: `Future(` for `str()` since `str()` needs to match schema syntax while `python_str` matches python